dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp

```diff
@@ -408,318 +407,114 @@ void DAGQueryBlockInterpreter::executeCastAfterTableScan(
-        {
-            match_helper_name = Join::match_helper_prefix + std::to_string(i);
-        }
+    RUNTIME_ASSERT(!input_streams_vec[1].empty(), log, "right input streams cannot be empty");
+    const Block & right_input_header = input_streams_vec[1].back()->getHeader();
 
-        columns_added_by_join.emplace_back(match_helper_name, Join::match_helper_type);
-        join_output_columns.emplace_back(match_helper_name, Join::match_helper_type);
-    }
-
-    DataTypes join_key_types;
-    getJoinKeyTypes(join, join_key_types);
-    TiDB::TiDBCollators collators;
-    size_t join_key_size = join_key_types.size();
-    if (join.probe_types_size() == static_cast<int>(join_key_size) && join.build_types_size() == join.probe_types_size())
-        for (size_t i = 0; i < join_key_size; i++)
-        {
-            if (removeNullable(join_key_types[i])->isString())
-            {
-                if (join.probe_types(i).collate() != join.build_types(i).collate())
-                    throw TiFlashException("Join with different collators on the join key", Errors::Coprocessor::BadRequest);
-                collators.push_back(getCollatorFromFieldType(join.probe_types(i)));
-            }
-            else
-                collators.push_back(nullptr);
-        }
-
-    Names left_key_names, right_key_names;
-    String left_filter_column_name, right_filter_column_name;
+    String match_helper_name = tiflash_join.genMatchHelperName(left_input_header, right_input_header);
+    NamesAndTypesList columns_added_by_join = tiflash_join.genColumnsAddedByJoin(build_pipeline.firstStream()->getHeader(), match_helper_name);
+    NamesAndTypes join_output_columns = tiflash_join.genJoinOutputColumns(left_input_header, right_input_header, match_helper_name);
 
     /// add necessary transformation if the join key is an expression
 
-    prepareJoin(
-        swap_join_side ? join.right_join_keys() : join.left_join_keys(),
-        join_key_types,
-        left_pipeline,
-        left_key_names,
+    bool is_tiflash_right_join = tiflash_join.isTiflashRightJoin();
+
+    // prepare probe side
+    auto [probe_side_prepare_actions, probe_key_names, probe_filter_column_name] = JoinInterpreterHelper::prepareJoin(
+        context,
+        probe_pipeline.firstStream()->getHeader(),
+        tiflash_join.getProbeJoinKeys(),
+        tiflash_join.join_key_types,
         true,
         is_tiflash_right_join,
-        swap_join_side ? join.right_conditions() : join.left_conditions(),
-        left_filter_column_name);
-
-    prepareJoin(
-        swap_join_side ? join.left_join_keys() : join.right_join_keys(),
-        join_key_types,
-        right_pipeline,
-        right_key_names,
+        tiflash_join.getProbeConditions());
+
+    // prepare build side
+    auto [build_side_prepare_actions, build_key_names, build_filter_column_name] = JoinInterpreterHelper::prepareJoin(
```

Comment:
I will add ` RUNTIME_ASSERT(probe_side_prepare_actions, log, "build_side_prepare_actions cannot be nullptr");` and `RUNTIME_ASSERT(probe_side_prepare_actions, log, "probe_side_prepare_actions cannot be nullptr");` later

Reply:
done